File: leetcode/palindrome-linked-list.cpp
#include <bits/stdc++.h>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* ptr = head;
        vector<int> list;
        
        while(ptr != nullptr) {
            list.push_back(ptr->val);
            ptr = ptr->next;
        }
        
        vector<int> reverse_list = list;
        reverse(reverse_list.begin(), reverse_list.end());
        
        return list == reverse_list;
    }
};

File: leetcode/palindrome-linked-list.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseLinkedList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        next = None

        if head:
            next = head.next
            head.next = None

        while next:
            next_ptr = next.next

            next.next = head
            head = next

            next = next_ptr

        return head

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        dummy = ListNode(self, head)

        slow_ptr = head
        fast_ptr = dummy

        while fast_ptr.next and fast_ptr.next.next:
            slow_ptr = slow_ptr.next
            fast_ptr = fast_ptr.next.next

        reversed_half = self.reverseLinkedList(slow_ptr)

        ptr = head
        reverse_ptr = reversed_half

        while ptr != slow_ptr and reverse_ptr:
            if ptr.val != reverse_ptr.val:
                return False

            ptr = ptr.next
            reverse_ptr = reverse_ptr.next

        return True
